Fix sigil formatting bug that produced invalid mixed syntax

williamthome opened the pull request:
Fixes #369

When formatting multiline sigil strings, erlfmt was incorrectly applying regular string transformation logic, creating invalid mixed syntax like:

```erlang
~"\n"
"    foo\n"
"    "
```

This fix leaves the sigil content entirely unchanged to prevent the mixed syntax bug.

I'm not entirely sure if this is the best solution - sigils might still be able to participate in container line-breaking while keeping their raw content intact.

I welcome maintainer feedback on whether full preservation fits erlfmt's formatting philosophy or if an alternative approach would be better.

williamthome committed:
Fix sigil formatting bug that produced invalid mixed syntax

When formatting multiline sigil strings, erlfmt was incorrectly applying regular string transformation logic, creating invalid mixed syntax like:

```erlang
~"\n"
"    foo\n"
"    "
```

This fix leaves the sigil content entirely unchanged to prevent the mixed syntax bug.

michalmuskala commented:
Sorry for the delay in reviewing this.

I'm fairly convinced this won't work correctly to indent the resulting strings inside sigils - the formatting algebra doesn't expect newlines inside strings - it expects to manage all newlines explicitly.

williamthome committed:
Improve sigil formatting to break containers with multiline sigils

When sigils contain newlines and are inside containers (lists, tuples,
function calls, maps, records), the container now properly breaks into
multiple lines while preserving the exact sigil content.

This addresses the formatting inconsistency where multiline sigils in
containers would not cause proper line breaks, leading to mixed
single-line and multi-line expressions.

Changes:
- Add force_breaks() to multiline sigils to signal parent containers
- Update tests to verify containers break properly with multiline sigils
- Preserve sigil binary content exactly (no indentation changes)

michalmuskala commented:
The only thing I'm still a bit apprehensive about is triple-quote strings - since then it's easy to properly handle newlines (and we already do it), I think for sigils with a triple-quote string inside we should do the previous code. They're not problematic since they strip-out the indentation whitespace properly. You can check out the snapshot tests in the `tripple_string.erl` - I think if we add a sigil to each one, they should behave the same as without it.

williamthome committed:
Refine sigil formatting to handle triple-quoted sigils properly

Following reviewer feedback, triple-quoted sigils now remain unchanged
in containers since they already handle whitespace and indentation
properly, unlike regular multiline sigils.

Changes:
- Split sigil handling into specific string pattern match and fallback
- Use same detection logic as string_to_algebra for consistency
- Triple-quoted sigils ("""...""") don't trigger force_breaks
- Regular multiline sigils still cause container breaking
- Added defensive fallback for non-string atomic types

This addresses the reviewer's concern about triple-quoted strings
handling indentation correctly and not needing container breaks.

williamthome committed:
Add comprehensive tests for triple-quoted sigil behavior

Expand test coverage to thoroughly verify that triple-quoted sigils
behave consistently across all container types and scenarios:

- Triple-quoted sigils with prefixes don't break containers
- Consistent behavior in tuples, function calls, and other contexts
- Mixed scenarios showing different behavior between triple-quoted
  and regular multiline sigils
- Edge cases with minimal content

This ensures the implementation correctly distinguishes between
sigil types and handles the reviewer's feedback about triple-quoted
strings managing indentation properly.

michalmuskala commented on src/erlfmt_format.erl:
```diff
@@ -259,9 +259,24 @@ do_expr_to_algebra({clauses, _Meta, Clauses}) ->
+        % Multi-line non-triple-quoted sigils need force_breaks
+        _ -> concat([force_breaks(), SigilDoc])
+    end;
+do_expr_to_algebra({sigil, _Meta, Prefix, {Atomic, ContentMeta, _Value}, Suffix}) when
```
Per the parser, sigil can only have string inside. This clause will never be executed.

https://github.com/WhatsApp/erlfmt/blob/1292f7c96b0d055f2e4c1facfdb54802e2ebdfbb/src/erlfmt_parse.yrl#L653

williamthome replied:
[08de55a](https://github.com/WhatsApp/erlfmt/pull/371/commits/08de55a5354dbddda7aeb754ac9363e7b979b0e7)

michalmuskala commented on src/erlfmt_format.erl:
```diff
@@ -259,9 +259,24 @@ do_expr_to_algebra({clauses, _Meta, Clauses}) ->
+    SigilDoc = concat(concat(PrefixDoc, string(Text)), do_expr_to_algebra(Suffix)),
+    case string:split(Text, "\n", all) of
+        % Triple-quoted sigils don't need force_breaks
+        ["\"\"\"" ++ _ | _] -> SigilDoc;
```
This will not work correctly for multiple nesting levels. We need to do the same logic as is done in normal triple-quoted strings in `string_to_algebra`, just add prefix and suffix around it.
Furthermore, ideally we'd avoid duplicating this logic and just share the code.

michalmuskala commented on test/erlfmt_format_SUITE.erl:
```diff
@@ -295,7 +295,126 @@ sigils(Config) when is_list(Config) ->
+    ?assertSame("~\"\"\"\nTest\nMultiline\n\"\"\"\n"),
+    %% Test multiline sigils in lists cause list to break
+    ?assertFormat(
+        "[~\"\n    foo\n    \"].\n",
```
For tests, please follow the existing code style and avoid `\n` directly in strings - split them across multiple lines. This makes it possible to easier understand what's going on from looking at the test - this is very hard to understand how things are moved around.

williamthome replied:
[d8cbea4](https://github.com/WhatsApp/erlfmt/pull/371/commits/d8cbea4f40e9f14148dcce4d08b010ef75f1e637)

williamthome committed:
Remove unreachable fallback clause from sigil handling

Per reviewer feedback, the parser only produces strings inside sigils,
so the fallback clause with ?IS_ATOMIC guard will never be executed.
Removed the dead code to clean up the implementation.

williamthome committed:
Improve test string formatting for better readability

Split multiline test strings across multiple concatenated string literals
instead of using \n escapes directly in strings. This makes the tests
much easier to read and understand what's happening.

Follows existing code style patterns in the test suite.

williamthome committed:
Fix triple-quoted sigils to break containers consistently

- Remove exemption for triple-quoted sigils from force_breaks logic
- All multiline sigils now consistently cause container breaking
- Update all triple-quoted sigil tests to expect proper container formatting
- Simplify logic to treat all multiline sigils uniformly

michalmuskala commented:
Alternative solution in https://github.com/WhatsApp/erlfmt/commit/5c1dc08573d44c74a507cd9912a65a0c9edb14bd that properly handles how we split things into newlines or not with just a final trailing newline
